cmd/rancher-desktop-guestagent/main.go

```diff
@@ -98,59 +97,61 @@ func main() {
 		cancel()
 	}()
 
-	tcpTracker := tcplistener.NewListenerTracker()
+	if !*enableContainerd && !*enableDocker {
+		log.Fatal("requires either -docker or -containerd enabled.")
+	}
 
-	wslAddr, err := getWSLAddr(wslInfName)
-	if err != nil {
-		log.Fatalf("failure getting WSL IP addresses: %v", err)
+	if *enableContainerd && *enableDocker {
+		log.Fatal("requires either -docker or -containerd, not both.")
 	}
 
-	forwarder := forwarder.NewVtunnelForwarder(*vtunnelAddr)
-	portTracker := tracker.NewPortTracker(forwarder, wslAddr)
+	var portTracker tracker.Tracker
 
 	if *enablePrivilegedService {
-		if !*enableContainerd && !*enableDocker {
-			log.Fatal("-privilegedService mode requires either -docker or -containerd enabled.")
-		}
-
-		if *enableContainerd && *enableDocker {
-			log.Fatal("-privilegedService mode requires either -docker or -containerd, not both.")
-		}
-
 		if *vtunnelAddr == "" {
 			log.Fatal("-vtunnelAddr must be provided when docker is enabled.")
 		}
 
-		if *enableContainerd {
-			group.Go(func() error {
-				eventMonitor, err := containerd.NewEventMonitor(*containerdSock, portTracker, tcpTracker)
-				if err != nil {
-					return fmt.Errorf("error initializing containerd event monitor: %w", err)
-				}
-				if err := tryConnectAPI(ctx, containerdSocketFile, eventMonitor.IsServing); err != nil {
-					return err
-				}
-				eventMonitor.MonitorPorts(ctx)
-
-				return eventMonitor.Close()
-			})
+		wslAddr, err := getWSLAddr(wslInfName)
+		if err != nil {
+			log.Fatalf("failure getting WSL IP addresses: %v", err)
 		}
 
-		if *enableDocker {
-			group.Go(func() error {
-				eventMonitor, err := docker.NewEventMonitor(portTracker)
-				if err != nil {
-					return fmt.Errorf("error initializing docker event monitor: %w", err)
-				}
-				if err := tryConnectAPI(ctx, dockerSocketFile, eventMonitor.Info); err != nil {
-					return err
-				}
-				eventMonitor.MonitorPorts(ctx)
-				eventMonitor.Flush()
-
-				return nil
-			})
-		}
+		forwarder := forwarder.NewVtunnelForwarder(*vtunnelAddr)
+		portTracker = tracker.NewPortTracker(forwarder, wslAddr)
+	} else {
+		portTracker = tracker.NewAPITracker(tracker.GatewayBaseURL)
+	}
+
+	if *enableContainerd {
+		group.Go(func() error {
+			eventMonitor, err := containerd.NewEventMonitor(*containerdSock, portTracker)
+			if err != nil {
+				return fmt.Errorf("error initializing containerd event monitor: %w", err)
+			}
+			if err := tryConnectAPI(ctx, containerdSocketFile, eventMonitor.IsServing); err != nil {
+				return err
+			}
+			eventMonitor.MonitorPorts(ctx)
+
+			return eventMonitor.Close()
+		})
+	}
+
+	if *enableDocker {
```

Comment:
Since `enableContainerd` and `enableDocker` are exclusive, should this be an `else if`? Or even drop one of them and just use `else`?

Reply:
I think having them separate provides better readability, `else` would not work since they can both be false, as for `else if` I tend to avoid complex conditions (for readability reasons).

Reply:
https://github.com/rancher-sandbox/rancher-desktop-agent/blob/977a193fd55f85f9a6b3685346291a56e873dff0/cmd/rancher-desktop-guestagent/main.go#L100-L102

We require one of the two to be true (as of this PR).

Reply:
This is now resolved — it's now possible to have them both enabled :D